design/kiota-e2e.md

```diff
@@ -0,0 +1,215 @@
+
+## Constraints
+
+Before we jump into the end-to-end walk-through, it's important to set some constraints.
+
+| Type      | Description                                                                                                                                   |
+| --------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
+| Platforms | Node : Current and Previous LTS (14 / 16)<br /> Web : Edge, Chrome, Firefox and Safari (Latest released version + immediate previous version) |
+| Modules   | Node : ESM (For preview) and potentially CJS (Based on feedback)<br /> Web : ESM                                                              |
+| Types     | Typings / Models should be available for both the core and the service libraries for Graph models                                             |
+
+## NodeJS e2e using the Service library
+
+### Concept
+
+This example is utilizing both the core and the service library. That means that as a developer, we want the full fluent API, allowing me discover APIs via the intellisence. Because the service library leverages the core library, we also want to highlight that we can leverage the core components directly (like the `.api()` method).
+
+### Install the Microsoft Graph Javascript Service Library
+
+There will be two versions of the service library for the Graph Javascript SDK. The default one will bring the representation of the v1.0 of Microsoft Graph. The second one will bring the representation of the beta of Microsoft Graph. This will be the only difference between these two libraries, as all the handcrafted code (delivering authentication, middlewares, tasks, etc.) will happen in the core library.
+
+The Microsoft Graph Javascript Service Library includes the following packages:
+
+-   `@microsoft/msgraph-sdk-javascript` ([npm](https://www.npmjs.com/package/@microsoft/msgraph-sdk-javascript-core)) - The service library for making fluent calls to Microsoft Graph v1.0.
+-   `@microsoft/msgraph-sdk-javascript-beta` ([npm](https://www.npmjs.com/package/@microsoft/msgraph-sdk-javascript-type)) - The service library for making fluent calls to Microsoft Graph beta.
+
+You can use [npm](https://www.npmjs.com) to install the Microsoft Graph Javascript Service Library :
+
+```Shell
+npm install @microsoft/msgraph-sdk-javascript --save
+npm install @microsoft/msgraph-sdk-javascript-beta --save-dev
+npm install @microsoft/kiota-authentication-azure --save
+```
+
+### Importing the right functionalities from the Graph Javascript Service Library
+
+```typescript
+// App.ts
+
+import { Client, User, Message, BodyType } from "@microsoft/msgraph-sdk-javascript";
+import { AzureIdentityAuthenticationProvider } from "@microsoft/kiota-authentication-azure";
```

Comment:
Ideally, we should avoid introducing yet another special way to do auth that works only in a particular case (in this case the MS Graph JS SDK). Instead, we should strive to use libraries like MSAL and @azure/identity that developers familiar with the Microsoft Identity Platform would already know. The fewer knew and specialistic concepts we introduce the lower the learning curve to start developing for Microsoft 365.

Reply:
this is just a wrapper around Azure identity, not a new auth lib

Reply:
Right, but developers don't know that and it requires us to create and maintain a separate set of samples, guidance, tutorials, etc and them to learn one way of doing auth with MSAL and another with our SDK. Is there a way for us to use what's already in MSAL so that we can have a consistent auth story across all Microsoft cloud?

Reply:
The decision to align on Azure Identity to provide consistency across Microsoft clouds (azure, office 365, etc...) was taken prior to Kiota work, almost 2 years ago now. In most cases, azure identity uses MSAL under the covers and adds value on top of it. The JavaScript SDK is the only "special" SDK in the sense that it might run in platforms that include some authentication mechanism (SPFx/TeamsFX). The current version of the SDK already supports Azure Identity in facts, https://github.com/microsoftgraph/msgraph-sdk-javascript/tree/dev/src/authentication/azureTokenCredentials

Reply:
And to expand on that, the reason why we didn't strip MSAL all together was because at the time azure identity was limited to node scenarios, which I don't believe is the case anymore.

Reply:
I agree, naming should be better here. If we were more specific with something like `@microsoft/kiota-provider-azure`. We would have the same for `@microsoft/kiota-provider-sharepoint`, `@microsoft/kiota-provider-teamsfx`, `@microsoft/kiota-provider-electron` allowing developer to focus on their business logic, doing the boilerplate code for them and them passing in the right context to allow seemless authentication (a TokenCredential for Azure Identity, a UserCredential for TeamsFx, a context for SharePoint, etc.)

Reply:
@sebastienlevert  do we have a confirmation that azure/identity works well for browser and we do not need an msal auth provider?

Also something to take a note of is that azure/identity does not provide a CDN source.

Reply:
@sebastienlevert if we're to provide providers for the different toolchains, what will that mean for the version matrix between the SDK and the different toolchains? What if we at Microsoft, or the community builds another toolchain: what will be the process to onboard it to Kiota to offer a consistent experience?

Reply:
This would not land in kiota. This could be a "community-led" kiota-supported authentication provider that anybody could leverage. It would not live in kiota (would leverage its set of interfaces as a dependency), but in a separate repo. For Microsoft, it could live in kiota if we believe there is a need / opportunity for us to deliver it. Right now, we are thinking to bring some of the MGT-specific providers (SPFx, TeamsFx, Electron) natively to kiota, so any kiota-built SDK would support these platforms also.

Reply:
Perhaps the consistency is not that relevant after all. Similarly how you can work with eslint plugins: some are provided by folks behind eslint, others come from the community. As long as there is a common contract and documentation how to build and set them up, their location and name might not matter.

> Right now, we are thinking to bring some of the MGT-specific providers (SPFx, TeamsFx, Electron) natively to kiota, so any kiota-built SDK would support these platforms also.

Do you have any example how this would be beneficial? Right now our toolchains are tied to a specific platform, eg. you can use SPFx, MGT or Teamsfx just with JavaScript, so what would be the benefit of having providers for these toolchains in Kiota if in the end they can work with just one provider and one Kiota SDK?

Reply:
Goal would be to use these providers with "any" Kiota-generated SDK. Any SDK could benefit from SPFx integration, same with TeamsFx for instance.